### PR TITLE
[Don't merge] Use conditional conformances to implement Equatable for Optional, Array and Dictionary

### PR DESCRIPTION
**Pull request by huonw**

### CODE REVIEW

File: stdlib/public/core/Arrays.swift.gyb [hunk 1]
```diff
@@ -2181,57 +2181,55 @@ extension _ArrayBufferProtocol {
...
+  }
 
-/// Returns `true` if the arrays do not contain the same elements.
-@_inlineable
-public func != <Element : Equatable>(
-  lhs: ${Self}<Element>, rhs: ${Self}<Element>
-) -> Bool {
-  return !(lhs == rhs)
+  /// Returns `true` if the arrays do not contain the same elements.
+  @_inlineable
```
Comment (slavapestov) [inline stdlib/public/core/Arrays.swift.gyb]: @airspeedswift We no longer need to define `operator!=` do we?

File: stdlib/public/core/Optional.swift [hunk 2]
```diff
@@ -313,98 +313,100 @@ func _diagnoseUnexpectedNilOptional(_filenameStart: Builtin.RawPointer,
...
+  ///     if numberToFind != numberFromString {
+  ///         print("No match.")
+  ///     }
+  ///     // Prints "No match."
+  ///
+  /// - Parameters:
+  ///   - lhs: An optional value to compare.
+  ///   - rhs: Another optional value to compare.
+  @_inlineable
+  public static func !=(lhs: Wrapped?, rhs: Wrapped?) -> Bool {
```
Comment (slavapestov) [inline stdlib/public/core/Optional.swift]: Ditto

File: stdlib/public/core/Arrays.swift.gyb (hunk 1 above)
Comment (moiseev) [reply stdlib/public/core/Arrays.swift.gyb]: Unless its implementation is more efficient than `!(==)`.

**Comment by moiseev** [comment]
@swift-ci Please smoke test

**Comment by moiseev** [comment]
@swift-ci please smoke test compiler performance

**Comment by DougGregor** [comment]
I'm taking this over in https://github.com/apple/swift/pull/12910, and it's really, really close

**Pull request closed without merging**